src/main/java/com/google/devtools/build/lib/skyframe/TreeArtifactValue.java

```diff
@@ -502,33 +571,18 @@ public interface TreeArtifactVisitor {
    * symlink that traverses outside of the tree artifact and any entry of {@link
    * Dirent.Type#UNKNOWN}, such as a named pipe.
    *
+   * <p>The visitor will be called on multiple threads in parallel. Accordingly, it must be
+   * thread-safe.
+   *
    * @throws IOException if there is any problem reading or validating outputs under the given tree
    *     artifact directory, or if {@link TreeArtifactVisitor#visit} throws {@link IOException}
    */
-  public static void visitTree(Path parentDir, TreeArtifactVisitor visitor) throws IOException {
-    visitTree(parentDir, PathFragment.EMPTY_FRAGMENT, checkNotNull(visitor));
-  }
-
-  private static void visitTree(Path parentDir, PathFragment subdir, TreeArtifactVisitor visitor)
-      throws IOException {
-    for (Dirent dirent : parentDir.getRelative(subdir).readdir(Symlinks.NOFOLLOW)) {
-      PathFragment parentRelativePath = subdir.getChild(dirent.getName());
-      Dirent.Type type = dirent.getType();
-
-      if (type == Dirent.Type.UNKNOWN) {
-        throw new IOException(
-            "Could not determine type of file for " + parentRelativePath + " under " + parentDir);
-      }
-
-      if (type == Dirent.Type.SYMLINK) {
-        checkSymlink(subdir, parentDir.getRelative(parentRelativePath));
-      }
-
-      visitor.visit(parentRelativePath, type);
-
-      if (type == Dirent.Type.DIRECTORY) {
-        visitTree(parentDir, parentRelativePath, visitor);
-      }
+  public static void visitTree(Path parentDir, TreeArtifactVisitor treeArtifactVisitor) throws IOException {
+    try {
+      Visitor visitor = new Visitor(parentDir, treeArtifactVisitor);
+      visitor.run();
+    } catch (InterruptedException e) {
+      throw new IOException(e);
```

Comment:
The key part that is different is really only this one compared to https://github.com/bazelbuild/bazel/commit/368bf117513731909337b6e2bd5082c046399506#commitcomment-110473970

Reply:
Yep, looks reasonable. Let's merge this!